fms/utils/generation.py

```diff
@@ -6,25 +6,36 @@
 import torch.nn.functional as F
 
 
-def __prepare_list_input(
-    input_ids_list: List[torch.Tensor], model: Union[Callable, torch.nn.Module]
+def make_padded(
+    input_ids_list: List[torch.Tensor], pad_to_length: Optional[int] = None
 ) -> Tuple[torch.Tensor, MutableMapping[str, Any]]:
     """
-    Convert the list of Tensors to a rectangular tensor. Return extra kwargs for the position_ids and mask, since this
-    will be required to properly handle the rectangular tensor for certain models.
+    Convert a list of Tensors to a rectangular tensor. Return extra padding kwargs for the position_ids and mask, since
+    this will be required to properly handle the rectangular tensor for certain models.
+
+    Parameters
+    ----------
+    input_ids_list: List[torch.Tensor]
+        a list of Tensors of varied length
+    pad_to_length: int, optional
+        If given, will pad to a specific max length. This value must be greater than or equal to the max length tensor.
+        (default is pad to pax tensor length)
+
+    Returns
+    -------
+    Tuple[torch.Tensor, MutableMapping[str, Any]]
+        A rectangular 2d padded tensor and a mapping containing the mask and position_ids typically used in forward pass
+        in fms models
+        A mapping from mask to a 3d causal mask and from position_ids to a 2d rectangular position_ids tensor
     """
-    min_len = min([seq.size(0) for seq in input_ids_list])
     max_len = max([seq.size(0) for seq in input_ids_list])
 
-    if isinstance(model, nn.Module):
-        forward_function = model.forward
-    else:
-        forward_function = model
-
-    params = inspect.signature(forward_function).parameters.keys()
-    extra_kwargs = {}
-    needs_mask = "mask" in params and min_len != max_len
-    needs_position_ids = "position_ids" in params and min_len != max_len
+    if pad_to_length is not None:
+        if pad_to_length < max_len:
+            raise ValueError(
+                f"pad_to_length (len={pad_to_length}) must be greater than or equal to the max length tensor (len={max_len})"
+            )
```

Comment:
I think it might be simpler to have the interface be something like `min_length = 0` instead of `pad_to_length = None`. The length can then always be max(min_length, *lengths) instead of needing to check for `if pad_to_length is not None`.

Does the user expect an error if they pass in a longer input_id? I'm not sure how the user would handle the ValueError besides truncating. maybe we could also offer a max_length and truncate?

Reply:
fixed this to `pad_to_min_length=0`. If `pad_to_min_length` is greater than the largest input_id (which is something we want to support), it will pad to that length, otherwise it will pad to the largest input_id.

Reply:
This is now `min_pad_length`